Widgets_CDAT/DictWidget/dict.py

```diff
@@ -0,0 +1,213 @@
+from PySide.QtCore import *
+from PySide.QtGui import *
+
+
+class InputChecker(QValidator):
+
+    inputInvalid = Signal()
+    correctInput = Signal()
+
+    def __init__(self, index):
+        super(InputChecker, self).__init__()
+        self.texts = []
+        self.index = index
+
+    def validate(self, input, pos):
+        if input in self.texts:
+            if self.texts.index(input) != self.index:
+                self.inputInvalid.emit()
+                return QValidator.Intermediate
+
+        self.correctInput.emit()
+        return QValidator.Acceptable
+
+    def setIndex(self, index):
+        self.index = index
+
+
+class DictEditor(QWidget):
+    textEdited = Signal(dict)
+
+    def __init__(self):
+        super(DictEditor, self).__init__()
+        self.valid_keys = None
+        self.keys = []
+        self.values = []
+        self.textEdits = []
+        self.validators = []
+        self.row_count = 0
+        self.rows = QVBoxLayout()
+        add_button = QPushButton()
+        add_button.setText("New Line")
+        add_button.clicked.connect(self.insertRow)
+
+        self.rows.addWidget(add_button)
+        self.setLayout(self.rows)
+
+    def colorInvalid(self, line_edit):
+        def applyColor():
+            line_edit.setStyleSheet("color: rgb(255, 0, 0);")
+        return applyColor
+
+    def colorValid(self, line_edit):
+        def applyColor():
+            line_edit.setStyleSheet("color: rgb(0, 0, 0);")
+        return applyColor
+
+
+    # Update Combo Boxes
+    def updateCBoxes(self, cur_box):
+        cur_text = cur_box.itemText(cur_box.currentIndex())
+        selectedItems = []
+        for box in self.cBoxes:
+            check_text = box.itemText(box.currentIndex())
+
+            selectedItems.append(check_text)
+
+            # check if boxes match and set to blank if so
+            if box != cur_box:
+                if check_text == cur_text:
+                    box.setCurrentIndex(0)
+
+        return selectedItems
+
+
+    # updates dict based on lineedit values
+    def updateEditTexts(self):
+        texts = []
+        for item in self.textEdits:
+            texts.append(item.text())
+        for i in self.validators:
+            i.texts = self.keys
+
+
+    # update lists and selectedItems
+    def updateLists(self, left, right):
+
+        def applyUpdate():
```

Comment:
I think it's time to teach you about partial functions; I ran into some "python is goofy" issues with the AdjustValues class related to these function factories.

```python
import functools

def myfunc(person, msg):
     print ("Hey %s!" % person), msg

bryce_func = functools.partial(myfunc, "Bryce")
bryce_func("This is a partial function!")
```

A partial function is a function that has some (or all) of the arguments assigned already, so you only need to pass in the stuff that you actually care about. It can be used in a very similar way to the function factory technique that I showed you, but it has slightly nicer behavior with regards to creating circular references (an object pointing at an object that points back at the original object), which garbage collectors have a hard time with. Try replacing these with partial function calls instead; so you'll just have

```python
def updateLists(self, left, right):
    ...
```

and in that function do all of the logic of `applyUpdate`.